src/ripple/app/tests/Env_test.cpp

```diff
@@ -0,0 +1,199 @@
+        env(clear(alice, asfDisableMaster), sig(bob));
+    }
+
+    // Multi-sign basics
+    void
+    testMultiSign()
+    {
+        using namespace jtx;
+
+        Env env(*this);
+        env.fund(XRP(10000), "alice");
+        env(signers("alice", 1,
+            { { "alice", 1 }, { "bob", 2 } }),                  ter(temBAD_SIGNER));
+        env(signers("alice", 1,
+            { { "bob", 1 }, { "carol", 2 } }));
+        env(noop("alice"));
+
+        env(noop("alice"), msig("bob"));
+        env(noop("alice"), msig("carol"));
+        env(noop("alice"), msig("bob", "carol"));
+        env(noop("alice"), msig("bob", "carol", "dilbert"),     ter(tefBAD_SIGNATURE));
+    }
+
+    // Two level Multi-sign
+    void
+    testMultiSign2()
+    {
+        using namespace jtx;
+
+        Env env(*this);
+        env.fund(XRP(10000), "alice", "bob", "carol");
+        env.fund(XRP(10000), "david", "eric", "frank", "greg");
+        env(signers("alice", 2, { { "bob", 1 },   { "carol", 1 } }));
+        env(signers("bob",   1, { { "david", 1 }, { "eric", 1 } }));
+        env(signers("carol", 1, { { "frank", 1 }, { "greg", 1 } }));
+
+        env(noop("alice"), msig2(
+        { { "bob", "david" } }),                                ter(tefBAD_QUORUM));
+        
+        env(noop("alice"), msig2(
+        { { "bob", "david" }, { "bob", "eric" } }),             ter(tefBAD_QUORUM));
+
+        env(noop("alice"), msig2(
+        { { "carol", "frank" } }),                              ter(tefBAD_QUORUM));
+        
+        env(noop("alice"), msig2(
+        { { "carol", "frank" }, { "carol", "greg" } }),         ter(tefBAD_QUORUM));
+
+        env(noop("alice"), msig2(
+        { { "bob", "david" }, { "carol", "frank" } }));
+
+        env(noop("alice"), msig2(
+        { { "bob", "david" }, { "bob", "eric" },
+          { "carol", "frank" }, { "carol", "greg" } }));
+    }
+
+    // Payment basics
+    void
+    testPayments()
+    {
+        using namespace jtx;
+
+        Env env(*this);
+        auto const gw = Account("gateway");
+        auto const USD = gw["USD"];
+
+        env(pay(env.master, "alice", XRP(1000)), fee(none),     ter(temMALFORMED));
+        env(pay(env.master, "alice", XRP(1000)), fee(1),        ter(telINSUF_FEE_P));
+        env(pay(env.master, "alice", XRP(1000)), seq(none),     ter(temMALFORMED));
+        env(pay(env.master, "alice", XRP(1000)), seq(2),        ter(terPRE_SEQ));
+        env(pay(env.master, "alice", XRP(1000)), sig(none),     ter(temMALFORMED));
+        env(pay(env.master, "alice", XRP(1000)), sig("bob"),    ter(tefBAD_AUTH_MASTER));
+
+        env(pay(env.master, "dilbert", XRP(1000)), sig(env.master));
+
+        env.fund(XRP(10000), "alice", "bob", "carol", gw);
+        expect(env["alice"].balance(XRP) == XRP(10000));
+        expect(env["bob"].balance(XRP) == XRP(10000));
+        expect(env["carol"].balance(XRP) == XRP(10000));
+        expect(env[gw].balance(XRP) == XRP(10000));
```

Comment:
Doesn't really matter, if the first 4 worked we can assume that the last one did. Remember, this is just to exercise the `Env` framework.